From ba1b600acb69ba4141a6ba2a908b9b75f40af4b0 Mon Sep 17 00:00:00 2001
From: Vignesh Sridharan <vignsrid@qti.qualcomm.com>
Date: Wed, 25 Mar 2026 15:15:56 +0530
Subject: [PATCH 1/1] config: Add SMD profile configurations to controller

Add support for up to 4 SMD (Seamless Mobility Domain) profiles in
controller configuration. Each profile includes identifier (MAC address),
SSID, timeout, max_prepared_ap_mlds, dl_forwarding, ptk_mode, type, and
default_dl_drain_time parameters.

Changes:
- Add SMD profile config example in beerocks_controller.conf.in
- Refactor config parsing to use loop-based approach for scalability
- Change sSmdProfileConfig::identifier from string to sMacAddr type
- Convert string identifier to sMacAddr during config parsing
- populate smd identifier in Agent AP MLD config TLV

Signed-off-by: Vignesh Sridharan <vignsrid@qti.qualcomm.com>
---
 .../bcl/include/bcl/beerocks_config_file.h    | 12 +++++
 .../bcl/source/beerocks_config_file.cpp       | 16 ++++++
 controller/config/beerocks_controller.conf.in | 15 ++++++
 .../beerocks/master/beerocks_master_main.cpp  | 50 +++++++++++++++++++
 controller/src/beerocks/master/controller.cpp | 24 +++++++++
 controller/src/beerocks/master/db/db.h        | 12 +++++
 6 files changed, 129 insertions(+)

diff --git a/common/beerocks/bcl/include/bcl/beerocks_config_file.h b/common/beerocks/bcl/include/bcl/beerocks_config_file.h
index b611a4b46..153c4e6bf 100644
--- a/common/beerocks/bcl/include/bcl/beerocks_config_file.h
+++ b/common/beerocks/bcl/include/bcl/beerocks_config_file.h
@@ -20,6 +20,8 @@ class config_file {
 public:
     typedef std::vector<std::tuple<std::string, std::string *, int>> tConfig;
 
+    static constexpr int MAX_SMD_PROFILES = 4;
+
     struct SConfigLog {
         //[log]
         std::string global_levels;
@@ -41,6 +43,16 @@ public:
         std::string ucc_listener_port;
         std::string load_legacy_client_roaming;
         std::string enable_smd;
+        // SMD Profile configurations (controller-side, indexed up to MAX_SMD_PROFILES)
+        std::string smd_profile_count;
+        std::string smd_identifier[MAX_SMD_PROFILES];
+        std::string smd_ssid[MAX_SMD_PROFILES];
+        std::string smd_timeout[MAX_SMD_PROFILES];
+        std::string smd_max_prepared_ap_mlds[MAX_SMD_PROFILES];
+        std::string smd_dl_forwarding[MAX_SMD_PROFILES];
+        std::string smd_ptk_mode[MAX_SMD_PROFILES];
+        std::string smd_type[MAX_SMD_PROFILES];
+        std::string smd_default_dl_drain_time[MAX_SMD_PROFILES];
         std::string load_service_fairness;
         std::string load_backhaul_measurements;
         std::string load_front_measurements;
diff --git a/common/beerocks/bcl/source/beerocks_config_file.cpp b/common/beerocks/bcl/source/beerocks_config_file.cpp
index 7994f9176..74c798953 100644
--- a/common/beerocks/bcl/source/beerocks_config_file.cpp
+++ b/common/beerocks/bcl/source/beerocks_config_file.cpp
@@ -54,6 +54,7 @@ bool config_file::read_master_config_file(const std::string &config_file_path, s
         std::make_tuple("load_legacy_client_roaming=", &conf.load_legacy_client_roaming,
                         mandatory_master),
         std::make_tuple("enable_smd=", &conf.enable_smd, 0),
+        std::make_tuple("smd_profile_count=", &conf.smd_profile_count, 0),
         std::make_tuple("load_service_fairness=", &conf.load_service_fairness, mandatory_master),
         std::make_tuple("load_backhaul_measurements=", &conf.load_backhaul_measurements, 0),
         std::make_tuple("load_front_measurements=", &conf.load_front_measurements, 0),
@@ -118,6 +119,21 @@ bool config_file::read_master_config_file(const std::string &config_file_path, s
         std::make_tuple("use_dataelements_vap_configs=", &conf.use_dataelements_vap_configs, 0),
     };
 
+    // Dynamically add SMD profile entries for each profile index up to MAX_SMD_PROFILES
+    for (int i = 0; i < config_file::MAX_SMD_PROFILES; i++) {
+        std::string prefix = "smd_" + std::to_string(i) + "_";
+        master_conf_args.emplace_back(prefix + "identifier=", &conf.smd_identifier[i], 0);
+        master_conf_args.emplace_back(prefix + "ssid=", &conf.smd_ssid[i], 0);
+        master_conf_args.emplace_back(prefix + "timeout=", &conf.smd_timeout[i], 0);
+        master_conf_args.emplace_back(
+            prefix + "max_prepared_ap_mlds=", &conf.smd_max_prepared_ap_mlds[i], 0);
+        master_conf_args.emplace_back(prefix + "dl_forwarding=", &conf.smd_dl_forwarding[i], 0);
+        master_conf_args.emplace_back(prefix + "ptk_mode=", &conf.smd_ptk_mode[i], 0);
+        master_conf_args.emplace_back(prefix + "type=", &conf.smd_type[i], 0);
+        master_conf_args.emplace_back(
+            prefix + "default_dl_drain_time=", &conf.smd_default_dl_drain_time[i], 0);
+    }
+
     bool ret_val = (read_config_file(config_file_path, master_conf_args, config_type) &&
                     read_log_section(config_file_path, conf.sLog));
 
diff --git a/controller/config/beerocks_controller.conf.in b/controller/config/beerocks_controller.conf.in
index 29203144c..defb47fda 100644
--- a/controller/config/beerocks_controller.conf.in
+++ b/controller/config/beerocks_controller.conf.in
@@ -18,6 +18,21 @@ load_legacy_client_roaming=1
 #   SMD feature:
 enable_smd=1 # 0 - disabled, 1 - enabled (SMD - Seamless Mobility Domain)
 
+# SMD Profile configurations (indexed, up to 4 profiles)
+# Each profile is identified by smd_identifier (MAC address)
+# Increment smd_profile_count to activate profiles
+smd_profile_count=0 # Number of active SMD profiles
+
+# Example SMD Profile 0 (uncomment and configure as needed):
+# smd_0_identifier=00:11:22:33:44:55  # Unique MAC address identifying this SMD
+# smd_0_ssid=MyNetwork                # SSID associated with this SMD profile
+# smd_0_timeout=100                   # ST preparation to execution timeout (TU units)
+# smd_0_max_prepared_ap_mlds=3        # Max number of prepared target AP MLDs
+# smd_0_dl_forwarding=1               # DL data forwarding support (1=yes, 0=no)
+# smd_0_ptk_mode=0                    # PTK mode: 0=per-SMD PTK, 1=per-AP PTK
+# smd_0_type=0                        # SMD type: 0=per-AP MLD MAC SAPs, 1=one MAC SAP
+# smd_0_default_dl_drain_time=10      # Default DL drain time (TU units)
+
 #   Measurement feature:
 load_backhaul_measurements=0
 load_front_measurements=1
diff --git a/controller/src/beerocks/master/beerocks_master_main.cpp b/controller/src/beerocks/master/beerocks_master_main.cpp
index 9fc1f2fc7..790378a9b 100644
--- a/controller/src/beerocks/master/beerocks_master_main.cpp
+++ b/controller/src/beerocks/master/beerocks_master_main.cpp
@@ -157,6 +157,56 @@ static void fill_master_config(son::db::sDbMasterConfig &master_conf,
     master_conf.load_service_fairness      = (main_master_conf.load_service_fairness == "1");
     master_conf.load_legacy_client_roaming = (main_master_conf.load_legacy_client_roaming == "1");
     master_conf.enable_smd                 = (main_master_conf.enable_smd == "1");
+    // Parse SMD profile configurations
+    {
+        int profile_count = 0;
+        if (!main_master_conf.smd_profile_count.empty()) {
+            profile_count = beerocks::string_utils::stoi(main_master_conf.smd_profile_count);
+        }
+        const int max_profiles = beerocks::config_file::MAX_SMD_PROFILES;
+        if (profile_count > max_profiles) {
+            LOG(WARNING) << "smd_profile_count=" << profile_count
+                         << " exceeds MAX_SMD_PROFILES=" << max_profiles << ", clamping to "
+                         << max_profiles;
+            profile_count = max_profiles;
+        }
+        for (int i = 0; i < profile_count; i++) {
+            son::db::sDbMasterConfig::sSmdProfileConfig profile;
+            profile.identifier           = main_master_conf.smd_identifier[i].empty()
+                                               ? beerocks::net::network_utils::ZERO_MAC
+                                               : tlvf::mac_from_string(main_master_conf.smd_identifier[i]);
+            profile.ssid                 = main_master_conf.smd_ssid[i];
+            profile.timeout              = main_master_conf.smd_timeout[i].empty()
+                                               ? 0
+                                               : static_cast<uint16_t>(beerocks::string_utils::stoi(
+                                        main_master_conf.smd_timeout[i]));
+            profile.max_prepared_ap_mlds = main_master_conf.smd_max_prepared_ap_mlds[i].empty()
+                                               ? 0
+                                               : static_cast<uint8_t>(beerocks::string_utils::stoi(
+                                                     main_master_conf.smd_max_prepared_ap_mlds[i]));
+            profile.dl_forwarding        = main_master_conf.smd_dl_forwarding[i].empty()
+                                               ? 0
+                                               : static_cast<uint8_t>(beerocks::string_utils::stoi(
+                                              main_master_conf.smd_dl_forwarding[i]));
+            profile.ptk_mode             = main_master_conf.smd_ptk_mode[i].empty()
+                                               ? 0
+                                               : static_cast<uint8_t>(beerocks::string_utils::stoi(
+                                         main_master_conf.smd_ptk_mode[i]));
+            profile.type                 = main_master_conf.smd_type[i].empty()
+                                               ? 0
+                                               : static_cast<uint8_t>(
+                                     beerocks::string_utils::stoi(main_master_conf.smd_type[i]));
+            profile.default_dl_drain_time =
+                main_master_conf.smd_default_dl_drain_time[i].empty()
+                    ? 0
+                    : static_cast<uint8_t>(beerocks::string_utils::stoi(
+                          main_master_conf.smd_default_dl_drain_time[i]));
+            LOG(DEBUG) << "Loaded SMD profile[" << i << "]: identifier=" << profile.identifier
+                       << ", ssid=" << profile.ssid << ", timeout=" << profile.timeout
+                       << ", max_prepared_ap_mlds=" << int(profile.max_prepared_ap_mlds);
+            master_conf.smd_profiles.push_back(profile);
+        }
+    }
     master_conf.load_backhaul_measurements = (main_master_conf.load_backhaul_measurements == "1");
     master_conf.load_front_measurements    = (main_master_conf.load_front_measurements == "1");
     master_conf.load_monitor_on_vaps       = (main_master_conf.load_monitor_on_vaps == "1");
diff --git a/controller/src/beerocks/master/controller.cpp b/controller/src/beerocks/master/controller.cpp
index 77992872d..7f3a81370 100644
--- a/controller/src/beerocks/master/controller.cpp
+++ b/controller/src/beerocks/master/controller.cpp
@@ -1423,6 +1423,18 @@ static bool add_agent_ap_mld_configuration_tlv(db &database, ieee1905_1::CmduMes
         ap_mld->modes().emlsr = mld.second.emlsr;
         ap_mld->modes().emlmr = mld.second.emlmr;
 
+        // Look up SMD profile by SSID and populate affiliated_ap_smd_mac_addr
+        ap_mld->modes().affiliated_ap_smd_mac_addr_valid = 0;
+        for (const auto &smd_profile : database.config.smd_profiles) {
+            if (!smd_profile.ssid.empty() && smd_profile.ssid == mld.second.ssid) {
+                ap_mld->affiliated_ap_smd_mac_addr()             = smd_profile.identifier;
+                ap_mld->modes().affiliated_ap_smd_mac_addr_valid = 1;
+                LOG(DEBUG) << "Set SMD MAC " << smd_profile.identifier << " for MLD SSID "
+                           << mld.second.ssid;
+                break;
+            }
+        }
+
         LOG(DEBUG) << "Creating AP MLD configuration for " << mld.second.ssid << "\n"
                    << "[ MLDID: " << mld.first << "]\n"
                    << "[ STR  : " << mld.second.str << "]\n"
@@ -4716,6 +4728,18 @@ bool Controller::reconfig_apmld(const sMacAddr &al_mac, const sMacAddr &mld_mac,
             ap_mld->modes().emlmr = 1;
         }
 
+        // Look up SMD profile by SSID and populate affiliated_ap_smd_mac_addr
+        ap_mld->modes().affiliated_ap_smd_mac_addr_valid = 0;
+        for (const auto &smd_profile : database.config.smd_profiles) {
+            if (!smd_profile.ssid.empty() && smd_profile.ssid == db_ap_mld.mld_info.mld_ssid) {
+                ap_mld->affiliated_ap_smd_mac_addr()             = smd_profile.identifier;
+                ap_mld->modes().affiliated_ap_smd_mac_addr_valid = 1;
+                LOG(DEBUG) << "Set SMD MAC " << smd_profile.identifier << " for MLD SSID "
+                           << db_ap_mld.mld_info.mld_ssid;
+                break;
+            }
+        }
+
         LOG(DEBUG) << "Sending AP MLD configuration for " << db_ap_mld.mld_info.mld_ssid
                    << " [mac=" << db_ap_mld.mld_info.mld_mac << ", mode=" << std::hex
                    << db_ap_mld.mld_info.mld_mode << "]";
diff --git a/controller/src/beerocks/master/db/db.h b/controller/src/beerocks/master/db/db.h
index a915b732b..bc151ce84 100644
--- a/controller/src/beerocks/master/db/db.h
+++ b/controller/src/beerocks/master/db/db.h
@@ -122,6 +122,18 @@ public:
         bool load_client_11k_roaming;
         bool load_legacy_client_roaming;
         bool enable_smd;
+        // SMD Profile configurations (controller-side, populated from config file)
+        struct sSmdProfileConfig {
+            sMacAddr identifier;               ///< SMD identifier as MAC address
+            std::string ssid;                  ///< SSID associated with this SMD profile
+            uint16_t timeout              = 0; ///< ST prep-to-execution timeout (TU units)
+            uint8_t max_prepared_ap_mlds  = 0; ///< Max number of prepared target AP MLDs
+            uint8_t dl_forwarding         = 0; ///< DL data forwarding support (1=yes, 0=no)
+            uint8_t ptk_mode              = 0; ///< PTK mode: 0=per-SMD PTK, 1=per-AP PTK
+            uint8_t type                  = 0; ///< SMD type: 0=per-AP MLD MAC SAPs, 1=one MAC SAP
+            uint8_t default_dl_drain_time = 0; ///< Default DL drain time (TU units)
+        };
+        std::vector<sSmdProfileConfig> smd_profiles;
         bool load_ire_roaming;
         bool load_load_balancing;
         bool load_service_fairness;
-- 
2.34.1

